packages/web-components/fast-router/src/enlistment.ts

```diff
@@ -0,0 +1,105 @@
+import { NavigationMessage, NavigationQueue } from "./navigation";
+import { NavigationTransaction } from "./transaction";
+import { Router } from "./router";
+import { RouterConfiguration } from "./configuration";
+import { LinkHandler } from "./links";
+
+export class MainEnlistment {
+    private navigationQueue: NavigationQueue | null = null;
+    private linkHandler: LinkHandler | null = null;
+    private config!: RouterConfiguration;
+    public readonly managesNavigation = true;
+
+    constructor(private router: Router) {
+        console.log("main enlistment", this);
+    }
+
+    public connect(config: RouterConfiguration) {
+        this.config = config;
+
+        if (this.navigationQueue !== null) {
+            this.navigationQueue.disconnect();
+        }
+
+        this.navigationQueue = config.createNavigationQueue();
+        this.navigationQueue.connect();
+        this.navigationQueue.receive().then(this.onNavigationMessage);
+
+        if (this.linkHandler !== null) {
+            this.linkHandler.disconnect();
+        }
+
+        this.linkHandler = this.config.createLinkHandler();
+        this.linkHandler.connect();
+    }
+
+    private onNavigationMessage = async (message: NavigationMessage) => {
+        const result = await this.config.findRoute(message.path);
+
+        if (result !== null) {
+            const transaction = new NavigationTransaction(
+                this.router,
+                this.config,
+                result.route,
+                result.command
+            );
+
+            await transaction.run();
+        }
+
+        this.navigationQueue!.receive().then(this.onNavigationMessage);
+    };
+}
+
+export class ChildEnlistment {
+    public readonly managesNavigation = false;
+
+    constructor(
+        private readonly parentRouter: Router,
+        private readonly childRouter: Router
+    ) {
+        console.log("child enlistment", this);
+        parentRouter.addNavigationParticipant(childRouter);
+    }
+}
+
+export type NavigationEnlistment = MainEnlistment | ChildEnlistment;
```

Comment:
Apparently `|` means union, is that what you're intending? Or should MainEnlistment and ChildEnlistment be considered separate for all types of NavigationEnlistment? That's why I created an `XOR` type in tooling see: https://github.com/microsoft/fast/blob/master/packages/tooling/fast-tooling/src/data-utilities/type.utilities.ts#L2